core/systemcontractupgrade/system-contract-upgrade.go

```diff
@@ -0,0 +1,116 @@
+package systemcontractupgrade
+
+import (
+	"encoding/hex"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+type UpgradeConfig struct {
+	BeforeUpgrade beforeUpgrade
+	AfterUpgrade  afterUpgrade
+	ContractAddr  common.Address
+	CommitUrl     string
+	Code          string
+}
+
+type Upgrade struct {
+	UpgradeName string
+	Configs     []*UpgradeConfig
+}
+
+type beforeUpgrade func(blockNumber *big.Int, contractAddr common.Address, statedb *state.StateDB)
+type afterUpgrade func(blockNumber *big.Int, contractAddr common.Address, statedb *state.StateDB)
+
+const (
+	mainnet = "Mainnet"
+	chapel  = "Chapel"
+	rialto  = "Rialto"
+)
+
+var (
+	// genesis contracts
+	validatorContract          = common.HexToAddress("0x0000000000000000000000000000000000001000")
```

Comment:
we have redefined the system contract address here, since parlia have done it too. But I don't know where to put it better either.

Reply:
System addresses in parlia.go are not completed. Besides, I think it would be better to defind all addresses in here ranther than in consensus engine.